### PR TITLE
Fix object reference leak from delegation.

### LINKED ISSUE #259: Object leak when using DelegatesTo and defaults.
With the following `EnabledAction` class, 4 gc-tracked objects are leaked on every `EnabledAction` instantiation: one each of types `Flag`, `list`, `CTrait` and `dict`.

```python
from traits.api import Bool, DelegatesTo, HasStrictTraits, Instance, Property


class Flag(HasStrictTraits):
    value = Bool


class EnabledAction(HasStrictTraits):
    visible = Property(Bool, depends_on=['can_enable', 'hide_if_restricted'])

    can_enable = DelegatesTo('can_enable_flag', prefix='value')

    can_enable_flag = Instance(Flag, kw={'value': True})

    hide_if_restricted = Bool(True)
```

Here's the full code showing the leak:
```python
from traits.api import Bool, DelegatesTo, HasStrictTraits, Instance, Property


class Flag(HasStrictTraits):
    value = Bool


class EnabledAction(HasStrictTraits):
    visible = Property(Bool, depends_on=['can_enable', 'hide_if_restricted'])

    can_enable = DelegatesTo('can_enable_flag', prefix='value')

    can_enable_flag = Instance(Flag, kw={'value': True})

    hide_if_restricted = Bool(True)


import cPickle
import collections
import gc

def detect_leaks():
    dumped_counts = []

    for _ in xrange(20):
        EnabledAction()
        gc.collect()
        objects = gc.get_objects()
        print len(objects)
        type_counts = collections.Counter(
            type(obj).__name__ for obj in objects)
        dumped_counts.append(cPickle.dumps(type_counts))
        del type_counts, objects

    counts = [
        cPickle.loads(dumped_count) for dumped_count in dumped_counts]

    print "Extra objects per run: ", counts[-1] - counts[-2]


if __name__ == '__main__':
    detect_leaks()
```

And the output from one run on my machine:
```
(.devenv)taniyama:Desktop mdickinson$ python testing.py 
14140
14161
14165
14169
14173
14177
14181
14185
14189
14193
14197
14201
14205
14209
14213
14217
14221
14225
14229
14233
Extra objects per run:  Counter({'list': 1, 'Flag': 1, 'CTrait': 1, 'dict': 1})
```

### PR DESCRIPTION
**Pull request by mdickinson**
Fixes #259.  ~~I'll add a regression test shortly.~~ Test added.

### CODE REVIEW

**Comment by mdickinson** [comment]
Appveyor is currently unhappy about building / installing babel on Python 3.x.  I'm not sure what's going on there.
```
Build started
$Env:sdkbin = "C:\Program Files\Microsoft SDKs\Windows\" + $Env:sdkver + "\Bin"
$Env:sdkverpath = "C:/Program Files/Microsoft SDKs/Windows/" + $Env:sdkver + "/Setup/WindowsSdkVer.exe"
$Env:path = $Env:python + ";" + $Env:python + "\scripts;" + $Env:sdkbin + ";" + $Env:path
Fetching repository commit (61f4117)...OK
Total: 1.8 MB in 277 files
Restoring build cache
Restoring "c:\egg_cache"...OK
Running Install scripts
if ((Test-Path "c:/egg_cache") -eq 0) { mkdir c:/egg_cache }
(new-object net.webclient).DownloadFile('https://raw.github.com/pypa/pip/master/contrib/get-pip.py', 'c:/get-pip.py')
python c:/get-pip.py
Requirement already up-to-date: pip in c:\python33-x64\lib\site-packages
Collecting wheel
  Downloading wheel-0.24.0-py2.py3-none-any.whl (63kB)
Installing collected packages: wheel
Successfully installed wheel-0.24.0
pip --version
pip 7.1.0 from C:\Python33-x64\lib\site-packages (python 3.3)
 
cmd /v:on /e:on /c ".\appveyor-install.cmd"
 
C:\projects\traits>"C:/Program Files/Microsoft SDKs/Windows/v7.1/Setup/WindowsSdkVer.exe" -q -version:"v7.1" 
 
C:\projects\traits>call setenv /x64 
Setting SDK environment relative to C:\Program Files\Microsoft SDKs\Windows\v7.1\.
Targeting Windows 7 x64 Debug
 
Collecting nose 
  Using cached nose-1.3.7-py3-none-any.whl 
Installing collected packages: nose 
Successfully installed nose-1.3.7 
Collecting coverage 
  Using cached coverage-3.7.1-cp33-none-win_amd64.whl 
Installing collected packages: coverage 
Successfully installed coverage-3.7.1 
Collecting numpy 
Installing collected packages: numpy 
Successfully installed numpy-1.9.2 
Collecting cython 
  Downloading Cython-0.23.tar.gz (1.6MB) 
Building wheels for collected packages: cython 
  Running setup.py bdist_wheel for cython 
  Stored in directory: c:/egg_cache\wheels\fa\8b\39\c26715a21c11bd43acb2bef754fdac0c028fa534108ac957dd 
Successfully built cython 
Installing collected packages: cython 
Successfully installed cython-0.23 
Collecting Sphinx 
  Using cached Sphinx-1.3.1-py2.py3-none-any.whl 
Collecting snowballstemmer>=1.1 (from Sphinx) 
Collecting Jinja2>=2.3 (from Sphinx) 
  Downloading Jinja2-2.8-py2.py3-none-any.whl (263kB) 
Collecting babel>=1.3 (from Sphinx) 
  Downloading Babel-2.0.tar.gz (3.4MB) 
Collecting alabaster<0.8,>=0.7 (from Sphinx) 
Collecting docutils>=0.11 (from Sphinx) 
  Using cached docutils-0.12-py3-none-any.whl 
Collecting Pygments>=2.0 (from Sphinx) 
  Using cached Pygments-2.0.2-py3-none-any.whl 
Collecting six>=1.4 (from Sphinx) 
  Using cached six-1.9.0-py2.py3-none-any.whl 
Collecting colorama (from Sphinx) 
Collecting sphinx-rtd-theme<0.2,>=0.1 (from Sphinx) 
  Using cached sphinx_rtd_theme-0.1.8-py2.py3-none-any.whl 
Collecting MarkupSafe (from Jinja2>=2.3->Sphinx) 
Collecting pytz>=0a (from babel>=1.3->Sphinx) 
  Using cached pytz-2015.4-py2.py3-none-any.whl 
Building wheels for collected packages: babel 
  Running setup.py bdist_wheel for babel 
  Stored in directory: C:/egg_cache\wheels\15\b9\ae\1b8f9a7c0e6e06a27d0bd8e6de2f3d82292c9b00e5d0a6dd45 
Successfully built babel 
Installing collected packages: snowballstemmer, MarkupSafe, Jinja2, pytz, babel, alabaster, docutils, Pygments, six, colorama, sphinx-rtd-theme, Sphinx 
Successfully installed Jinja2-2.8 MarkupSafe-0.23 Pygments-2.0.2 Sphinx-1.3.1 alabaster-0.7.6 babel-2.0 colorama-0.3.3 docutils-0.12 pytz-2015.4 six-1.9.0 snowballstemmer-1.2.0 sphinx-rtd-theme-0.1.8 
Traceback (most recent call last):
  File "setup.py", line 160, in <module>
    cmdclass=additional_commands(),
  File "setup.py", line 115, in additional_commands
    from sphinx.setup_command import BuildDoc
  File "C:\Python33-x64\lib\site-packages\sphinx\setup_command.py", line 23, in <module>
    from sphinx.application import Sphinx
  File "C:\Python33-x64\lib\site-packages\sphinx\application.py", line 37, in <module>
    from sphinx.builders import BUILTIN_BUILDERS
  File "C:\Python33-x64\lib\site-packages\sphinx\builders\__init__.py", line 23, in <module>
    from sphinx.util import i18n, path_stabilize
  File "C:\Python33-x64\lib\site-packages\sphinx\util\i18n.py", line 15, in <module>
    from babel.messages.pofile import read_po
  File "C:\Python33-x64\lib\site-packages\babel\messages\__init__.py", line 12, in <module>
    from babel.messages.catalog import *
  File "C:\Python33-x64\lib\site-packages\babel\messages\catalog.py", line 23, in <module>
    from babel.dates import format_datetime
  File "C:\Python33-x64\lib\site-packages\babel\dates.py", line 28, in <module>
    from babel.util import UTC, LOCALTZ
  File "C:\Python33-x64\lib\site-packages\babel\util.py", line 278, in <module>
    from babel import localtime
  File "C:\Python33-x64\lib\site-packages\babel\localtime\__init__.py", line 21, in <module>
    from babel.localtime._win32 import _get_localzone
  File "C:\Python33-x64\lib\site-packages\babel\localtime\_win32.py", line 18, in <module>
    tz_names = get_global('windows_zone_mapping')
  File "C:\Python33-x64\lib\site-packages\babel\core.py", line 58, in get_global
    _global_data = pickle.load(fileobj)
TypeError: an integer is required (got type str)
Command exited with code 1
```

**Comment by mdickinson** [comment]
Something that still bothers me a bit is that that regression test doesn't fail reliably on master: if I set `PYTHONHASHSEED=random`, the test sometimes passes, sometimes fails.  It's not clear to me where the dictionary-order dependence is coming from in this case.

**Comment by mdickinson** [comment]
Anyone want to review?  I'd welcome a second pair of eyes to verify that I haven't changed the semantics (apart from the refcount fix) in `ctraits.c`.

**Comment by itziakos** [comment]
:+1:

File: traits/ctraits.c [hunk 1]
```diff
@@ -1147,16 +1147,26 @@ _has_traits_trait ( has_traits_object * obj, PyObject * args ) {
...
-        if ( ((dict == NULL) ||
-              ((temp_delegate = (has_traits_object *) PyDict_GetItem( dict,
-                                          trait->delegate_name )) == NULL)) &&
-              ((temp_delegate = (has_traits_object *) has_traits_getattro(
-                  delegate, trait->delegate_name )) == NULL) )
-            break;
 
+        temp_delegate = NULL;
+        if (dict != NULL) {
+            temp_delegate = (has_traits_object *) PyDict_GetItem(
```
Comment (mdickinson) [inline traits/ctraits.c]: Hmm.  A `NULL` return here indicates that an exception was set.  We should be either clearing the exception or propagating it down the call stack.  As it is, the code just ignores it and continues with the next `if`.

**Comment by mdickinson** [comment]
There's something still not quite right with this code (both pre- and post- this PR).  See comment above.

File: traits/ctraits.c (hunk 1 above)
Comment (mdickinson) [inline traits/ctraits.c]: Fixed.

Comment (mdickinson) [inline traits/ctraits.c]: Nope, that's the wrong thing to do:  `PyDict_GetItem` does *not* set an exception for a missing key; it merely returns `None`.

Comment (mdickinson) [inline traits/ctraits.c]: "fix" reverted.  I think the code's okay as it is.

**Comment by mdickinson** [comment]
@itziakos: Want to sign off on the latest changes?

**Comment by itziakos** [comment]
:+1:

**Comment by mdickinson** [comment]
Thanks!  Merging.

**Pull request merged**